File: Managed/Core/Rendering/Light.cs
using DX12Demo.Core.Binding;
using Newtonsoft.Json;
using System.Numerics;

namespace DX12Demo.Core.Rendering
{
    public enum LightType : int
    {
        Directional = 0, // 平行光
        Point = 1,       // 点光源
        Spot = 2,        // 聚光灯
    };

    public partial class Light : Component
    {
        private nint m_Light;

        public Light()
        {
            m_Light = Light_New();
        }

        ~Light()
        {
            Dispose();
        }

        [JsonProperty] public LightType Type
        {
            get => Light_GetType(m_Light);
            set => Light_SetType(m_Light, value);
        }

        [JsonProperty] public Color Color
        {
            get => Light_GetColor(m_Light);
            set => Light_SetColor(m_Light, value);
        }

        [JsonProperty] public Vector2 FalloffRange
        {
            get => Light_GetFalloffRange(m_Light);
            set => Light_SetFalloffRange(m_Light, value);
        }

        [JsonProperty] public float SpotPower
        {
            get => Light_GetSpotPower(m_Light);
            set => Light_SetSpotPower(m_Light, value);
        }

        private void Dispose()
        {
            if (m_Light == nint.Zero)
            {
                return;
            }

            Light_Delete(m_Light);
            m_Light = nint.Zero;
        }

        protected override void OnMount()
        {
            base.OnMount();
            RenderPipeline.AddLight(m_Light);
        }

        protected override void OnUnmount()
        {
            RenderPipeline.RemoveLight(m_Light);
            Dispose();
            GC.SuppressFinalize(this);
            base.OnUnmount();
        }

        protected override void OnEnable()
        {
            base.OnEnable();
            Light_SetIsActive(m_Light, true);
        }

        protected override void OnDisable()
        {
            Light_SetIsActive(m_Light, false);
            base.OnDisable();
        }

        protected override void OnUpdate()
        {
            base.OnUpdate();

            Light_SetPosition(m_Light, MountingObject.Position);
            Light_SetRotation(m_Light, MountingObject.Rotation);
        }

        [NativeFunction]
        private static partial nint Light_New();

        [NativeFunction]
        private static partial void Light_Delete(nint self);

        [NativeFunction]
        private static partial void Light_SetPosition(nint self, Vector3 value);

        [NativeFunction]
        private static partial void Light_SetRotation(nint self, Quaternion value);

        [NativeFunction]
        private static partial void Light_SetIsActive(nint self, bool value);

        [NativeFunction]
        private static partial LightType Light_GetType(nint self);

        [NativeFunction]
        private static partial void Light_SetType(nint self, LightType value);

        [NativeFunction]
        private static partial Color Light_GetColor(nint self);

        [NativeFunction]
        private static partial void Light_SetColor(nint self, Color value);

        [NativeFunction]
        private static partial Vector2 Light_GetFalloffRange(nint self);

        [NativeFunction]
        private static partial void Light_SetFalloffRange(nint self, Vector2 value);

        [NativeFunction]
        private static partial float Light_GetSpotPower(nint self);

        [NativeFunction]
        private static partial void Light_SetSpotPower(nint self, float value);
    }
}

File: Source/Native/RenderPipeline/Private/Light.cpp
#include "Light.h"
#include "Transform.h"
#include "RenderPipeline.h"
#include "Application.h"

using namespace DirectX;

namespace march
{
    Light::Light()
        : Type(LightType::Directional)
        , Color(1.0f, 1.0f, 1.0f, 1.0f)
        , FalloffRange(1.0f, 10.0f)
        , SpotPower(64.0f)
    {
    }

    void Light::FillLightData(LightData& data) const
    {
        if (Type == LightType::Directional)
        {
            XMFLOAT3 forward = GetTransform()->GetForward();
            data.Position.x = -forward.x;
            data.Position.y = -forward.y;
            data.Position.z = -forward.z;
            data.Position.w = 0;
        }
        else
        {
            data.Position.x = GetTransform()->GetPosition().x;
            data.Position.y = GetTransform()->GetPosition().y;
            data.Position.z = GetTransform()->GetPosition().z;
            data.Position.w = 1;
        }

        if (Type == LightType::Spot)
        {
            XMFLOAT3 forward = GetTransform()->GetForward();
            data.SpotDirection.x = -forward.x;
            data.SpotDirection.y = -forward.y;
            data.SpotDirection.z = -forward.z;
            data.SpotDirection.w = SpotPower;
        }
        else
        {
            data.SpotDirection.x = 0;
            data.SpotDirection.y = 0;
            data.SpotDirection.z = 0;
            data.SpotDirection.w = 0;
        }

        data.Color.x = Color.x;
        data.Color.y = Color.y;
        data.Color.z = Color.z;
        data.Color.w = 1;

        data.Falloff.x = FalloffRange.x;
        data.Falloff.y = FalloffRange.y;
        data.Falloff.z = 0;
        data.Falloff.w = 0;
    }

    void Light::OnMount()
    {
        Component::OnMount();
        GetApp()->GetRenderPipeline()->AddLight(this);
    }

    void Light::OnUnmount()
    {
        GetApp()->GetRenderPipeline()->RemoveLight(this);
        Component::OnUnmount();
    }
}
